### PR TITLE
EPMLABSBRN-414 #Registration

### PR DESCRIPTION
**Pull request by SiestaVie**
[EPMLABSBRN-414](https://jira.epam.com/jira/browse/EPMLABSBRN-414)

**Description**:new ExceptionHandler was added and bean-validaator for messageSource. Test was added and editconfig as well.
Txt files was deleted

### CODE REVIEW

File: words_series1.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+бал,бум,быль,вить,гад,дуб,линь,лис,моль,пар,пять,раб,рак,рожь,сеть,топь,ход,шеф
```
Comment (goopnigoop) [inline words_series1.txt:1]: Just a question why do we have these files (words_series1.txt, words_series2.txt) as a part of this commit?

Comment (SiestaVie) [reply words_series1.txt:1]: I guess there is code for this files which creates them. We could put it in .gitignore file

Comment (ElenaSpb) [reply words_series1.txt:1]: this files are need to generation audio files tool, so they can be merged now I suppose it  might be useful until cration audiofiles logic is not implemented

Comment (goopnigoop) [reply words_series1.txt:1]: @ElenaSpb our features should include only files that belong to the feature, you can not put there anything else.
These files should be added as a separate task.
@SiestaVie please, don't include these, thank you

File: src/main/resources/messages_en.properties [hunk 2]
```diff
@@ -1,4 +1,6 @@
 group.first.name=non-speech exercises
 group.second.name=speech exercises
 group.first.description=non-speech exercises
-group.second.description=speech exercises
\ No newline at end of file
+group.second.description=speech exercises
```
Comment (mike050210) [inline src/main/resources/messages_en.properties:4]: @lifeart was mentioning that we might not need to handle different localizations in BE as they are doing it from the FE. @lifeart, If we return a map like @SiestaVie, I think it won't matter which language is the actual message as she is also returning the field names that failed the validations. What do you think?

Comment (lifeart) [reply src/main/resources/messages_en.properties:4]: correct, error may looks like

{
  name: REGISTRATION_DUPLICATED_FIELD,
  path: 'login'
}

and on FE we will map `REGISTRATION_DUPLICATED_FIELD` to selected language, and `path` for special field

File: words_series1.txt (hunk 1 above)
Comment (ElenaSpb) [reply words_series1.txt:1]: yes, logically this files should be added in separate MR

File: src/main/kotlin/com/epam/brn/controller/advice/ExceptionControllerAdvice.kt [hunk 3]
```diff
@@ -81,8 +86,32 @@ class ExceptionControllerAdvice {
...
+    }
+
+    @ExceptionHandler(MethodArgumentNotValidException::class)
+    fun methodArgumentNotValidExceptionHandler(
+        ex: MethodArgumentNotValidException
+    ): ResponseEntity<ApiError> {
+        val errors = HashMap<String, ArrayList<String>>()
+        val listName = ArrayList<String>()
+        val listPath = ArrayList<String>()
+        for (violation in ex.bindingResult.allErrors) {
```
Comment (mike050210) [inline src/main/kotlin/com/epam/brn/controller/advice/ExceptionControllerAdvice.kt:101]: You can also return `fieldErrors` directly and avoid the condition below

File: api-contract/samples/errors_for_ui [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# REGISTRATION ----------------------
+firstName = First Name must not be blank
```
Comment (ElenaSpb) [inline api-contract/samples/errors_for_ui:2]: as it is constants, lets name it like
FIRST_NAME_VALIDATION_MESSAGE
LAST_NAME_VALIDATION_MESSAGE
EMAIL_NAME_VALIDATION_MESSAGE

File: src/main/kotlin/com/epam/brn/dto/UserAccountDto.kt [hunk 5]
```diff
@@ -10,9 +10,9 @@ import javax.validation.constraints.Size
 @JsonInclude(JsonInclude.Include.NON_NULL)
 data class UserAccountDto(
     val id: Long? = null,
-    @field:NotBlank
+    @field:NotBlank(message = "First Name must not be blank")
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/UserAccountDto.kt:13]: and there should be only the key like
FIRST_NAME_VALIDATION_MESSAGE

File: src/main/kotlin/com/epam/brn/dto/UserAccountDto.kt [hunk 6]
```diff
@@ -10,9 +10,9 @@ import javax.validation.constraints.Size
...
     val firstName: String,
-    @field:NotBlank
+    @field:NotBlank(message = "Last Name must not be blank")
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/UserAccountDto.kt:15]: and there should be only the key like
LAST_NAME_VALIDATION_MESSAGE

File: src/main/kotlin/com/epam/brn/controller/advice/ExceptionControllerAdvice.kt [hunk 7]
```diff
@@ -81,8 +86,32 @@ class ExceptionControllerAdvice {
...
+            .contentType(MediaType.APPLICATION_JSON)
+            .body(BaseResponseDto(errors = listOf(e.message.toString())))
+    }
+
+    @ExceptionHandler(MethodArgumentNotValidException::class)
+    fun methodArgumentNotValidExceptionHandler(
+        ex: MethodArgumentNotValidException
+    ): ResponseEntity<ApiError> {
+        val errors = HashMap<String, ArrayList<String>>()
+        val listName = ArrayList<String>()
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/controller/advice/ExceptionControllerAdvice.kt:99]: lets put to errors only keys - names

**Comment by ElenaSpb** [comment]
and please put example of response into raml api
like
registrationBadResponse.json:
_{
  "data": [],
  "errors": [
    "FIRST_NAME_MUST_NOT_HAVE_SPACES", "LAST_NAME_MUST_NOT_HAVE_SPACES","EMAIL_MUST_NOT_HAVE_SPACES"
  ],
  "meta": []
}_

File: src/main/kotlin/com/epam/brn/config/WebMvcBasicConfiguration.kt [hunk 8]
```diff
@@ -22,4 +25,22 @@ class WebMvcBasicConfiguration : WebMvcConfigurer {
...
+    }
+
+    @Bean
+    fun messageSource(): ReloadableResourceBundleMessageSource? {
+        val messageSource = ReloadableResourceBundleMessageSource()
+        messageSource.setBasenames(
+            "classpath:/messages",
+            "classpath:authErrorResponse"
+        )
+        messageSource.setDefaultEncoding("UTF-8")
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/config/WebMvcBasicConfiguration.kt:43]: could you please use const for UTF-8 like
StandardCharsets.UTF_8.name()

File: src/main/kotlin/com/epam/brn/controller/advice/ExceptionControllerAdvice.kt [hunk 9]
```diff
@@ -81,8 +85,27 @@ class ExceptionControllerAdvice {
...
+            .body(BaseResponseDto(errors = listOf(e.message.toString())))
+    }
+
+    @ExceptionHandler(MethodArgumentNotValidException::class)
+    fun methodArgumentNotValidExceptionHandler(
+        ex: MethodArgumentNotValidException
+    ): ResponseEntity<ApiError> {
+        val errors = ArrayList<String>()
+        for (violation in ex.bindingResult.allErrors) {
+            if (violation is FieldError) {
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/controller/advice/ExceptionControllerAdvice.kt:99]: we do not use {} for single code line in if-else bloks
https://github.com/Brain-up/brn/wiki/Code-Style

File: src/main/kotlin/com/epam/brn/dto/BaseResponseDto.kt [hunk 10]
```diff
@@ -11,3 +13,5 @@ data class BaseSingleObjectResponseDto(
     val errors: List<Any> = emptyList(),
     val meta: List<Any> = emptyList()
 )
+
+data class ApiError(val errors: ArrayList<String>)
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/BaseResponseDto.kt:17]: why we  need this class? 
we can delete it I suppose now

File: src/main/resources/authErrorResponse.properties [hunk 11]
```diff
@@ -0,0 +1,3 @@
+first.name.not.spaces=FIRST_NAME_MUST_NOT_HAVE_SPACES
```
Comment (ElenaSpb) [inline src/main/resources/authErrorResponse.properties:1]: lets change not -> without

File: src/main/kotlin/com/epam/brn/controller/advice/ExceptionControllerAdvice.kt [hunk 12]
```diff
@@ -81,8 +85,27 @@ class ExceptionControllerAdvice {
...
-                    .status(HttpStatus.INTERNAL_SERVER_ERROR)
-                    .contentType(MediaType.APPLICATION_JSON)
-                    .body(BaseResponseDto(errors = listOf(e.message.toString())))
+            .status(HttpStatus.INTERNAL_SERVER_ERROR)
+            .contentType(MediaType.APPLICATION_JSON)
+            .body(BaseResponseDto(errors = listOf(e.message.toString())))
+    }
+
+    @ExceptionHandler(MethodArgumentNotValidException::class)
+    fun methodArgumentNotValidExceptionHandler(
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/controller/advice/ExceptionControllerAdvice.kt:94]: could you please name this functions as other in this class
handleMethodArgumentNotValidException

File: src/test/kotlin/com/epam/brn/controller/advice/ExceptionControllerAdviceTest.kt [hunk 13]
```diff
@@ -60,5 +71,28 @@ internal class ExceptionControllerAdviceTest {
         assertTrue((responseEntity.body as BaseResponseDto).errors.toString().contains("some test exception"))
         assertEquals(HttpStatus.INTERNAL_SERVER_ERROR, responseEntity.statusCode)
         assertEquals(MediaType.APPLICATION_JSON, responseEntity.getHeaders().getContentType())
-        }
+    }
+
+    @Test
+    fun `should make MethodArgumentNotValidException`() {
```
Comment (ElenaSpb) [inline src/test/kotlin/com/epam/brn/controller/advice/ExceptionControllerAdviceTest.kt:77]: why `make`?
maybe `handle`?

File: src/main/resources/messages_en.properties (hunk 2 above)
Comment (ElenaSpb) [reply src/main/resources/messages_en.properties]: ok, lets have response like 
{
  "data": [],
  "errors": [
    "FIRST_NAME_MUST_NOT_HAVE_SPACES", "LAST_NAME_MUST_NOT_HAVE_SPACES","EMAIL_MUST_NOT_HAVE_SPACES"
  ],
  "meta": []
}

File: src/main/kotlin/com/epam/brn/config/WebMvcBasicConfiguration.kt [hunk 14]
```diff
@@ -22,4 +26,22 @@ class WebMvcBasicConfiguration : WebMvcConfigurer {
...
+
+    @Bean
+    fun validator(): Validator? {
+        val factory = LocalValidatorFactoryBean()
+        factory.setValidationMessageSource(messageSource()!!)
+        return factory
+    }
+
+    @Bean
+    fun messageSource(): ReloadableResourceBundleMessageSource? {
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/config/WebMvcBasicConfiguration.kt:38]: why the type is `ReloadableResourceBundleMessageSource?`
it should be `ReloadableResourceBundleMessageSource`

File: src/main/kotlin/com/epam/brn/config/WebMvcBasicConfiguration.kt [hunk 15]
```diff
@@ -22,4 +26,22 @@ class WebMvcBasicConfiguration : WebMvcConfigurer {
     override fun addInterceptors(registry: InterceptorRegistry) {
         registry.addInterceptor(LocaleChangeInterceptor())
     }
+
+    @Bean
+    fun validator(): Validator? {
+        val factory = LocalValidatorFactoryBean()
+        factory.setValidationMessageSource(messageSource()!!)
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/config/WebMvcBasicConfiguration.kt:33]: we can try to avoid !!

**Pull request merged**